drivers: counter: update esp32 support

sylvioalves opened the pull request:
Update ESP32 counter driver to:
- Add reset support
- Add top value callback support
- Add guard period support
- Add reliable cancel test support

```c
------ TESTSUITE SUMMARY START ------
SUITE PASS - 100.00% [counter_basic]: pass = 9, fail = 0, skip = 1, total = 10 duration = 4.287 seconds
 - PASS - [counter_basic.test_all_channels] duration = 0.261 seconds
 - PASS - [counter_basic.test_cancelled_alarm_does_not_expire] duration = 2.050 seconds
 - PASS - [counter_basic.test_late_alarm] duration = 0.203 seconds
 - PASS - [counter_basic.test_late_alarm_error] duration = 0.203 seconds
 - SKIP - [counter_basic.test_multiple_alarms] duration = 0.203 seconds
 - PASS - [counter_basic.test_set_top_value_with_alarm] duration = 0.312 seconds
 - PASS - [counter_basic.test_short_relative_alarm] duration = 0.205 seconds
 - PASS - [counter_basic.test_single_shot_alarm_notop] duration = 0.343 seconds
 - PASS - [counter_basic.test_single_shot_alarm_top] duration = 0.273 seconds
 - PASS - [counter_basic.test_valid_function_without_alarm] duration = 0.234 seconds
SUITE PASS - 100.00% [counter_no_callback]: pass = 1, fail = 0, skip = 0, total = 1 duration = 0.206 seconds
 - PASS - [counter_no_callback.test_set_top_value_without_alarm] duration = 0.206 seconds
------ TESTSUITE SUMMARY END ------
===================================================================
PROJECT EXECUTION SUCCESSFUL
```

sylvioalves committed:
drivers: counter: esp32: add reset counter function

Adds option to reset counter value.

Signed-off-by: Sylvio Alves <sylvio.alves@espressif.com>

sylvioalves committed:
drivers: counter: esp32: add top value handling

Implements top value callback and handling.

Signed-off-by: Sylvio Alves <sylvio.alves@espressif.com>

sylvioalves committed:
drivers: counter: esp32: add guard period support

Implements guard period support.

Signed-off-by: Sylvio Alves <sylvio.alves@espressif.com>

sylvioalves committed:
tests: counter: add esp32 timer in cancel test case

Allow ESP32 devices to test the scenario where
when alarm should not be triggered after being cancelled.

Signed-off-by: Sylvio Alves <sylvio.alves@espressif.com>

joelguittet commented:
@sylvioalves other MCUs have a counter entry in the timer, for example https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/arm/st/l4/stm32l4.dtsi#L320
Possible to have the same for ESP32 so then drivers using counter feature are effectively portable?

joelguittet commented:
@sylvioalves for the latest proposal regarding the counter node in the device tree, I just created a PR at https://github.com/zephyrproject-rtos/zephyr/pull/90300.

sylvioalves committed:
drivers: counter: esp32: remove rtc counter test

As part of counter API updates, RTC timer will be removed from
counter API driver and can be removed from the test list.

Signed-off-by: Sylvio Alves <sylvio.alves@espressif.com>
